File: src/GnOuGo.Flow.Core/Runtime/Executors/SwitchExecutor.cs
using System.Text.Json.Nodes;
using GnOuGo.Flow.Core.Expressions;
using GnOuGo.Flow.Core.Models;

namespace GnOuGo.Flow.Core.Runtime.Executors;

/// <summary>
/// Switch step: evaluates cases and executes the first match.
/// Form A: expr + cases with value matches
/// Form B: cases with when conditions
/// </summary>
public sealed class SwitchExecutor : IStepExecutor
{
    public StepRecovery Recovery => StepRecovery.Composite;
    public string StepType => "switch";

    public IReadOnlyList<StepExceptionDoc>? DocumentedExceptions => new StepExceptionDoc[]
    {
        new(ErrorCodes.InputValidation, false, "The step is missing `cases` or exceeds the runtime maximum number of switch cases.")
    };

    public string DslSnippet => """
        ### switch — Conditional branching
        Form A (value match):
        ```yaml
        - id: route
          type: switch
          expr: "${data.inputs.mode}"
          cases:
            - value: "fast"
              steps:
                - id: fast_step
                  type: template.render
                  input: { engine: mustache, template: "Fast mode", mode: text }
            - value: "slow"
              steps: [...]
          default:
            - id: default_step
              type: template.render
              input: { engine: mustache, template: "Default", mode: text }
        ```
        Form B (when expressions):
        ```yaml
        - id: route
          type: switch
          cases:
            - when: "${data.inputs.score > 80}"
              steps: [...]
            - when: "${data.inputs.score > 50}"
              steps: [...]
          default: [...]
        ```
        Output: output of the matched branch's steps.
        """;

    public async Task<JsonNode?> ExecuteAsync(StepExecutionContext ctx, CancellationToken ct)
    {
        var cases = ctx.Step.Cases;
        if (cases == null || cases.Count == 0)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation, "switch step requires 'cases'");

        if (cases.Count > ctx.Limits.MaxSwitchCases)
            throw new WorkflowRuntimeException(ErrorCodes.InputValidation,
                $"Switch cases ({cases.Count}) exceeds limit ({ctx.Limits.MaxSwitchCases})");

        var selected = await ctx.RecordControlAsync("branch", () =>
        {
            var expression = ctx.Step.Source.Expr is null ? null : ctx.Interpolator.Interpolate(ctx.Step.Source.Expr, ctx.Data);
            for (var index = 0; index < cases.Count; index++)
            {
                var item = cases[index];
                var match = expression is not null && item.Source.Value is not null
                    ? item.Source.Value == (expression is JsonValue scalar && scalar.TryGetValue(out string? text) ? text : expression.ToJsonString())
                    : item.Source.When is not null && ExpressionEvaluator.GetBool(ctx.Interpolator.Interpolate(item.Source.When, ctx.Data));
                if (match) return JsonValue.Create(index);
            }
            return JsonValue.Create(-1);
        }, ct);
        var branch = selected!.GetValue<int>();
        if (branch >= 0)
        {
            var result = new RunResult { Success = true };
            await ctx.Engine.ExecuteStepsAsync(cases[branch].Steps, ctx.Data, result, ctx.Limits, ctx.CallDepth, ctx.CallStack,
                ctx.EffectiveExecutionScope.Child("case", branch.ToString(System.Globalization.CultureInfo.InvariantCulture)), ct, ctx.TelemetrySpan);
            return ctx.Data["steps"]?.DeepClone();
        }

        // Default branch
        if (ctx.Step.Default != null && ctx.Step.Default.Count > 0)
        {
            var result = new RunResult { Success = true };
            await ctx.Engine.ExecuteStepsAsync(
                ctx.Step.Default,
                ctx.Data,
                result,
                ctx.Limits,
                ctx.CallDepth,
                ctx.CallStack,
                ctx.EffectiveExecutionScope.Child("case", "default"),
                ct,
                ctx.TelemetrySpan);
            return ctx.Data["steps"]?.DeepClone();
        }

        return null;
    }
}

File: librairies/python/gnougo-flow-core/src/gnougo_flow_core/runtime_steps/switch_executor.py
from __future__ import annotations

import json
from typing import Any

from gnougo_flow_core.runtime import *  # noqa: F401,F403


def _expr_to_string(value: Any) -> str:
    """Mirror .NET: JsonValue.s if string, else ToJsonString()."""
    if isinstance(value, str):
        return value
    if value is True:
        return "true"
    if value is False:
        return "false"
    if value is None:
        return "null"
    try:
        return json.dumps(value, ensure_ascii=False, default=str)
    except Exception:
        return str(value)


class SwitchExecutor:
    step_type = "switch"
    step_description = "Branch using expr/value or case conditions."
    dsl_snippet = """
### switch - Branch execution by value or condition
```yaml
- id: route
  type: switch
  expr: "${data.inputs.mode}"
  cases:
    - value: quick
      steps:
        - id: quick_path
          type: set
          input: { selected: quick }
    - when: "${data.inputs.priority > 5}"
      steps:
        - id: priority_path
          type: set
          input: { selected: priority }
  default:
    - id: fallback
      type: set
      input: { selected: default }
```
Output: merged `data.steps` after selected branch execution.
"""
    documented_exceptions = [
        (
            ErrorCodes.INPUT_VALIDATION,
            False,
            "The step is missing `cases` or exceeds the runtime maximum number of switch cases.",
        ),
    ]

    async def execute_async(self, ctx: StepExecutionContext) -> Any:
        if not ctx.step.cases:
            raise WorkflowRuntimeException(ErrorCodes.INPUT_VALIDATION, "switch step requires 'cases'")
        if len(ctx.step.cases) > ctx.limits.max_switch_cases:
            raise WorkflowRuntimeException(
                ErrorCodes.INPUT_VALIDATION,
                f"Switch cases ({len(ctx.step.cases)}) exceeds limit ({ctx.limits.max_switch_cases})",
            )

        # Form A: evaluate expr once
        expr_value: Any = None
        if ctx.step.source.expr is not None:
            expr_value = ctx.engine.interpolator.interpolate(ctx.step.source.expr, ctx.data)

        for case in ctx.step.cases:
            matched = False

            # Form A: value match — both expr and case value must be non-null (mirrors .NET)
            if expr_value is not None and case.source.value is not None:
                expr_str = _expr_to_string(expr_value)
                matched = case.source.value == expr_str
            # Form B: when condition (also acts as Form A fallthrough when value is null)
            elif case.source.when is not None:
                cond = ctx.engine.interpolator.interpolate(case.source.when, ctx.data)
                matched = ExpressionEvaluator.get_bool(cond)

            if matched:
                run = RunResult(success=True)
                await ctx.engine.execute_steps_async(
                    case.steps, ctx.data, run, ctx.limits, ctx.call_depth, ctx.call_stack, ctx.telemetry_span, ct=ctx.ct,
                )
                return dict(ctx.data.get("steps", {}))

        if ctx.step.default:
            run = RunResult(success=True)
            await ctx.engine.execute_steps_async(
                ctx.step.default, ctx.data, run, ctx.limits, ctx.call_depth, ctx.call_stack, ctx.telemetry_span, ct=ctx.ct,
            )
            return dict(ctx.data.get("steps", {}))
        return None
